model/TotpPf.php

```diff
@@ -0,0 +1,419 @@
+<?php
+use OTPHP\TOTP;
+use Endroid\QrCode\Builder\Builder;
+use Endroid\QrCode\Encoding\Encoding;
+use Endroid\QrCode\ErrorCorrectionLevel\ErrorCorrectionLevelHigh;
+use Endroid\QrCode\Label\Alignment\LabelAlignmentCenter;
+use Endroid\QrCode\Label\Font\NotoSans;
+use Endroid\QrCode\RoundBlockSizeMode\RoundBlockSizeModeMargin;
+use Endroid\QrCode\Writer\PngWriter;
+
+class TotpPf
+{
+	private $key_table;
+	private $table;
+
+	public function __construct(string $tableName)
+	{
+		$ok = ['mailbox', 'admin'];
+
+		if (!in_array($tableName, $ok)) {
+			throw new \InvalidArgumentException("Unsupported tableName for TOTP: " . $tableName);
+		}
+		$this->table = $tableName;
+		$this->key_table = table_by_key($tableName);
+		$this->login = new Login($tableName);
+	}
+
+	/**
+	 * @param string &$qr_code for returning base64-encoded qr-code
+	 *
+	 * @return string TOTP_secret, empty if NULL
+	 * @throws \Exception if invalid user, or db update fails.
+	 */
+	public function generate($username, &$qr_code): string
```

Comment:
can this change to return an array like ['totp_secret' => .... , 'qr_code' => ....] ?

I dislike passing by reference for some reason

Reply:
Old habits. 
Changed.